README.md
# RLDS
Reinfrocement Learning for Dialouge Stragtegy


NLU_Table/history.txt
    1  ls
    2  lspci -vnn | grep VGA -A 12
    3  lshw 
    4  lshw -numeric -C display
    5  python
    6  l
    7  ls
    8  cd ..
    9  ls
   10  python
   11  ls
   12  cat examples.desktop 
   13  Following is the example to demonstrate InputStream and OutputStream −
   14  import java.io.*;
   15  public class fileStreamTest {
   16  }
   17  The above code would create file test.txt and would write given numbers in binary format. Same would be the output on the stdout screen.
   18  File Navigation and I/O
   19  There are several other classes that we would be going through to get to know the basics of File Navigation and I/O.
   20  Directories in Java
   21  A directory is a File which can contain a list of other files and directories. You use File object to create directories, to list down files available in a directory. For complete detail, check a list of all the methods which you can call on File object and what are related to directories.
   22  Creating Directories
   23  There are two useful File utility methods, which can be used to create directories −
   24  Following example creates "/tmp/user/java/bin" directory −
   25  Example
   26  import java.io.File;
   27  public class CreateDir {
   28  }
   29  Compile and execute the above code to create "/tmp/user/java/bin".
   30  Note − Java automatically takes care of path separators on UNIX and Windows as per conventions. If you use a forward slash (/) on a Windows version of Java, the path will still resolve correctly.
   31  Listing Directories
   32  You can use list( ) method provided by File object to list down all the files and directories available in a directory as follows −
   33  Example
   34  import java.io.File;
   35  public class ReadDir {
   36  }
   37  This will produce the following result based on the directories and files available in your /tmp directory −
   38  Output
   39  test1.txt
   40  test2.txt
   41  ReadDir.java
   42  ReadDir.class
   43  Previous Page
   44  Next Page  
   45  whois
   46  who
   47  whatis
   48  who
   49  whois
   50  uname -r
   51  lsb_release 
   52  lsb_release -a
   53  apt-get install vncserver
   54  sudo 
   55  sudo su
   56  apt-get download bpython
   57  sudo apt-get install bpython
   58  apt-get download bpython
   59  ls
   60  dpkg -i bpython_0.15-2_all.deb --force-not-root --root=$HOME
   61  apt-get source package
   62  apt-get source bpython
   63  source ~/.bashrc 
   64  python
   65  curl -O https://repo.continuum.io/archive/Anaconda3-4.3.1-Linux-x86_64.sh
   66  ls
   67  sha256sum Anaconda3-4.3.1-Linux-x86_64.sh 
   68  bash Anaconda3-4.3.1-Linux-x86_64.sh 
   69  ls
   70  rm -rf anaconda3/
   71  ls
   72  ls -a
   73  md5sum Anaconda2-5.1.0-Linux-x86_64.sh 
   74  man md5sum
   75  md5sum Anaconda2-5.1.0-Linux-x86_64.sh -c  5b1b5784cae93cf696e11e66983d8756
   76  man md5sum
   77  md5sum Anaconda2-5.1.0-Linux-x86_64.sh -t  5b1b5784cae93cf696e11e66983d8756
   78  nano anacondamd5.txt
   79  md5sum Anaconda2-5.1.0-Linux-x86_64.sh -c anacondamd5.txt 
   80  bash Anaconda2-5.1.0-Linux-x86_64.sh 
   81  rm -rf anaconda2/
   82  ls
   83  bash Anaconda2-5.1.0-Linux-x86_64.sh 
   84  ls
   85  source ~/.bashrc
   86  nano .bashrc
   87  conda list
   88  conda --help
   89  conda update -n base conda
   90  source activate tensorlow
   91  source activate tensorflow
   92  python
   93  clear
   94  python
   95  ldconfig -v
   96  ldconfig -v | grep cuda
   97  ldconfig -v | grep libcublas
   98  export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/usr/local/cuda/extras/CUPTI/lib64
   99  python
  100  export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/usr/local/cuda/extras/CUPTI/lib64
  101  history
  102  pip install --ignore-installed --upgrade https://storage.googleapis.com/tensorflow/linux/gpu/tensorflow_gpu-1.6.0-cp27-none-linux_x86_64.whl
  103  python
  104  ldconfig -v | grep libcublas
  105  ldconfig -v 
  106  l
  107  nano .bashrc
  108  source ~/.bashrc
  109  source ./bashrc
  110  source ~/.bashrc
  111  source activate tensorflow
  112  python
  113  source deactiavte
  114  source deactivate
  115  history
  116  python
  117  python 
  118  ls
  119  history 
  120  lsb_release -a
  121  uname -r
  122  uname
  123  uname -a
  124  lshw
  125  cat /proc/cpuinfo
  126  python 
  127  free -m
  128  free -g
  129  dpkg -i teamviewer_13.0.9865_amd64.deb 
  130  remmina -h
  131  ifconfig
  132  wwho
  133  who
  134  ifconfig
  135  exit
  136  python
  137  wo
  138  who
  139  wocat /usr/local/cuda/include/cudnn.h | grep CUDNN_MAJOR -A 2
  140  cat /usr/local/cuda/include/cudnn.h | grep CUDNN_MAJOR -A 2
  141  cat /usr/local/cuda
  142  ls /usr/local/cuda
  143  cat /usr/local/cuda/README 
  144  cat /usr/local/cuda/version.txt 
  145  pip install tensorflow-gpu
  146  python
  147  pip uninstall tensorflow
  148  ls
  149  pip install --ignore-installed --upgrade tf_nightly_gpu-1.head-cp27-none-linux_x86_64.whl 
  150  python
  151  pip install keras
  152  python
  153  cd RLforDialouge-1/
  154  ls
  155  cd TulikaMaam/
  156  l
  157  ls
  158  python train_modified.py 
  159  pip install gym
  160  python train_modified.py 
  161  who]
  162  who
  163  nvidia-smi
  164  watch -n 0.5 nvidia-smi
  165  watch -n -d 0.5 nvidia-smi
  166  watch -d -n 0.5 nvidia-smi
  167  nvidia-smi -l 1
  168  vidia-settings -q GPUUtilization -q useddedicatedgpumemory
  169  nvidia-settings -q GPUUtilization -q useddedicatedgpumemory
  170  nvidia-settings
  171  nvidia-settings  -h
  172  nvidia-settings -q all
  173  top
  174  htop
  175  nvidia-smi
  176  lshw
  177  lshw -C display
  178  lspci
  179  lspci | grep nvidia
  180  lspci | grep gpu
  181  lspci | grep Inte
  182  lspci | grep Nv
  183  lspci -v
  184  exit
  185  ls
  186  cd ..
  187  cd TulikaMaam/
  188  python
  189  python train_modified.py 
  190  export CUDA_VISIBLE_DEVICES=''
  191  python train_modified.py 
  192  ls
  193  cd RLforDialouge-1/
  194  ls
  195  cd TulikaMaam/

NLU_Deep/env1.py
# -*- coding: utf-8 -*-
"""
Created on Fri Dec  8 06:40:03 2017

@author: dawg
Last File : envOld2
Changes made:
1. Going to implement the action taken counter
2. Giving reward for greeting in the starting with negative reward
3. 
all changes with comment @1/3/18

Updated after @6/3/18

Previous file : env.py picking up from tulika maams work

Things to be implemented
1. the positive reward to 1
2. Cap for the negative reward
    2,1 w2*threshold*5
    2.2 w3
3. Added the significa  of w2 again
previous file is envOld2.py
"""

import numpy as np
import random
from functions1 import *


class DialougeSimulation(object):
    """
    State Descipriton:
    State: contains 6 variables (greet, deptCity, arrCity, deptTime, depDay, class ,number of iteration uptil no)
     Action Set :
     greet (0), ask{arrCity,deptCity,deptTime,depDay,class}()
     reask/confirm{arrCity,deptCity,deptTime,depDay,class} .
     askDeptandArr, askDateTime # hybrid actions

    # adding a seveth parameter for number of iterations
    """

    # @1/3/18 changing the default weight value of w2 to 8 and w3 to 13
    def __init__(self, w1=1, w2=8, w3=13):
        """
        The weights are as follows
        w1 : is the interaction weigt
        w2 : is the change in confidence value
        w3: is the weight given to weight value
        """
        self.state_size = 5
        self.actions = 13
        self.current_state = []  # the current state of the agent
        self.states = np.zeros([0, 5])  # this is the collection of all the states
        self.init_state()
        self.num_iter = 0
        # self.actions = ['greet','askDeptCity','askArrCity','askDepDay','askDepTime','askClass','askDeptArrCity','askDateTIme','reaskDeptCity','reaskArrCity','reaskDepDay','reaskDepTime','reaskClass','closeConversation' ]
        self.w1 = w1  # interaction weight
        self.w2 = w2
        self.w3 = w3
        # 1/3/18 Keep a track of the actions taken i.e a count of number of action taken for each type
        self.actions_taken = np.zeros([self.actions])
        # @6/3/18 calculate the max possible value
        self.threshold = 0.7
        #self.max_reward = max(self.w3, self.w2 * 5 * self.threshold)

    def init_state(self):
        self.states = np.array([[0.0, 0.0, 0.0, 0.0, 0.0]])  # initliase with zero for all values
        self.current_state = self.states[-1]
        # 1/3/18 reinit the actions taken matrix
        self.actions_taken = np.zeros([self.actions])

    def random_init_state(self):
        self.states = np.array(
            [[random.random() for i in range(self.state_size)]])  # initialise with zero for all values
        self.current_state = self.states[-1]
        # 1/3/18 reinit the actions taken matrix
        self.actions_taken = np.zeros([self.actions])


    def check(self, action):
        self.threshold = 0.7
        if (self.threshold < self.current_state[action - 7 - 1]):
            return 1
        else:
            return 0

    def step(self, action):
        # this will be the next step to take for an action lies between 0 - 12 including both
        """
        Now the confidence value of the slots will be defined on the language unit of the system:
        We will assume the langugae system to be a random funciton generator which
        """
        done = False

        reward = 0
        # action is a numeric value

        new_state = np.array([float('%.4f' % elem) for elem in self.current_state])
        #if action == 0 and self.num_iter == 0:  # greeting should be first
            #new_state[0] = 1
        if action >= 0 and action <= 4:  # the action is to ask a slot value
            new_state[action] = 0.2 * random.random() + 0.55  # the confidenec value lies between 0.6 - 0.8
        if action > 6 and action <= 11:
            # this will seq increase the confidence value action 8 - 7
            # @1/3/18 18:49:00 Update : reask will only assign confidence once the slot has been asked for
            if new_state[action - 7] < 0.1:  # @1/3/18 18:49:00this implies that the ask part has not happened
                pass
            else:
                new_state[action - 7] = (1 - new_state[action - 7]) * 0.85 + new_state[
                    action - 7]  # @1/3/18 18:49:00the ask part has happned
        if action == 5:
            new_state[0] = 0.2 * random.random() + 0.55
            new_state[1] = 0.2 * random.random() + 0.55
        if action == 6:
            new_state[2] = 0.2 * random.random() + 0.55
            new_state[3] = 0.2 * random.random() + 0.55

        # @1/3/18 Increase the action taken value for the action taken 
        # self.actions_taken[action] += 1

        #if action == 0 and self.num_iter == 0:  # reward for greeting
            # @1/3/18 adding a negative reward for not greeting
            #reward = self.w3
        #elif self.num_iter == 0 and action != 0:
            #print("w3=", self.w3)
            #reward = -self.w3  # @1/3/18 giving the negative reward for not greeting
        #elif self.num_iter != 0 and action == 0:
            #reward = -self.w3
        if action == 12:
            done = True  # set the sim finished
            # 28/2/18 impleted the soft check functionalt
            val = soft_check_state(self.current_state)

            if val >= 0:
                """
                The returned value will be negative if less than 3 states are satifies
                """
                # @7/3/18 added w2 again
                reward = val * (sum(self.current_state))*self.w2  # 1/3/18 removing the greeting from final reward + self.w3*self.current_state[0])
                reward = abs(reward)
                print("r in if=" + str(reward))
            else:
                # @7/3/18 added w2 again
                reward = -self.w2*(sum(np.full(5, 1)) - sum(self.current_state))  # 1/3/18 removing the reward from final greeting + self.w3*self.current_state[0]
                print("r in else=" + str(reward))
        else:
            # @1/3/18 we have to update the self.w1 to be affected byt he action taken 
            print(new_state)
            print(self.current_state)
            # @7/3/18 added w2 again
            reward = self.w2*(sum(new_state) - sum(self.current_state))  # - self.w1*self.actions_taken[action] # @1/3/18 a factor of action_taken is given
            #print(reward)
            #print(self.actions_taken[action])
            reward = reward - self.w1 
            # val = new_check_state(self.current_state)
            # print("val=" + str(val))
            # reward = reward + val


        self.actions_taken[action] += 1
        # 28/2/18 add a noise to the state
        noise_tobe_added = np.random.rand(len(self.current_state))*0.02 -0.01
        # 28/2/18 now we will remove noise from greeting and iterations
        # noise_tobe_added[0] = 0 # greet state to be zero
        # noise_tobe_added[-1] = 0 # iteratiion state to be zero
        self.num_iter = self.num_iter + 1
        #new_state[6] = self.num_iter
        self.current_state = np.array([float('%.3f' % elem) for elem in new_state])

        # adding the noise to the current state
        # self.current_state = self.current_state + noise_tobe_added

        self.states = np.append(self.states, [self.current_state], axis=0)
        reward = float(reward)#/self.max_reward

        return self.current_state, reward, done  # return the state and reward
        # at the moment we will consider the end episode to return the reward

    # need an indentifier to indetify the end of DialougeSimualtion
    def reset(self):
        """
        reset the state to the statrting poitn
        """
        i = random.random()
        if i < 0.5:
            self.init_state()
        else:
            self.random_init_state()
        self.num_iter = 0

        return self.current_state
